### PR TITLE
Adds support for managing Azure DevOps service connections and their permissions

### PR DESCRIPTION
**Pull request by JamesDawson**
Updated Cmdlets:
- `Assert-RequiredResourceAccessContains` - Fixes issue with the specified type of assignment not being honoured
- `Connect-Azure` - Improves validation of existing connections

New Cmdlets:
- `Assert-AzdoServiceConnection`
- `Assert-ResourceGroupWithRbac`
- `Assert-AzureAdApiPermissions`
- `Assert-AzureServicePrincipalForRbac`
- `Assert-AzureCliExtension`
- `Get-MicrosoftGraphApiAppUri`

### CODE REVIEW

File: module/functions/azdo/Assert-AzdoServiceConnection.ps1 [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+    if (!$existingAdoServiceConnection) {
+        Write-Host "A new ADO service connection will be created"
+        $existingSp,$spSecret = Assert-AzureServicePrincipalForRbac -Name $ServicePrincipalName `
+                                                                    -WhatIf:$WhatIfPreference
+        
+        # check we have the secret for the SPN
+        if (!$spSecret) {
+            if ($PSCmdlet.ShouldProcess($Name, "Reset Service Principal Credential")) {
+                Write-Warning "The service principal already exists, but we do not have the secret - to proceed, its password must be reset"
```
Comment (MikeLarah) [inline module/functions/azdo/Assert-AzdoServiceConnection.ps1:93]: Do we want to reset the credential here, or could we just add a new secret? What if other processes are using the existing secret?

File: module/functions/azure/aad/Assert-AzureServicePrincipalForRbac.ps1 [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            Write-Host ("Complete - ApplicationId={1}" -f $Name, $newSp.appId)
+            $spSecret = $newSp.password
+        
+            $checkNewSpArgs = @(
+                "ad sp list"
+                "--display-name $Name"
+            )
+            $existingSp = Invoke-AzCli $checkNewSpArgs -asJson
+            if (!$existingSp) {
+                # TODO: retry?
```
Comment (MikeLarah) [inline module/functions/azure/aad/Assert-AzureServicePrincipalForRbac.ps1:63]: Intentional TODO left here?

File: module/functions/azdo/Assert-AzdoServiceConnection.ps1 (hunk 1 above)
Comment (JamesDawson) [reply module/functions/azdo/Assert-AzdoServiceConnection.ps1:93]: I know what you mean, but these SPNs are supposed to be dedicated to this purpose and if they do exist then they might be getting a bunch of new permissions, so arguably there's a case for revoking old credentials that previous users/processes might have known.

File: module/functions/azure/aad/Assert-AzureServicePrincipalForRbac.ps1 (hunk 2 above)
Comment (JamesDawson) [reply module/functions/azure/aad/Assert-AzureServicePrincipalForRbac.ps1:63]: Kind of, I was wondering whether we might see latency issues where the newly created SP is not immediately visible.

File: module/functions/azdo/Assert-AzdoServiceConnection.ps1 (hunk 1 above)
Comment (JamesDawson) [reply module/functions/azdo/Assert-AzdoServiceConnection.ps1]: This behaviour is now behind a switch.  Scenarios running without the switch that require the secret to be reset will be blocked.

**Review by MikeLarah — APPROVED** [review]

**Pull request merged**